### PR TITLE
Start of issue 718, file not existing clean up

### PR DESCRIPTION
**Pull request by jessicamizzi**
Fix for #718.

### CODE REVIEW

**Comment by jessicamizzi** [comment]
- [ ] Is it mergable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage.
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

**Comment by jessicamizzi** [comment]
Can 2 people please review this code? @b-wyss @bocajnotnef @brtaylor92

**Comment by bocajnotnef** [comment]
Looks good to me minus the non-mergable status.

**Comment by b-wyss** [comment]
Looks good to me as well, unless you want to change existance to existence? Totally up to you, but the rest looks okay!

File: khmer/kfile.py [hunk 1]
```diff
@@ -22,7 +22,16 @@ def check_file_status(file_path, force):
 
     if file_path is '-':
         return
-    mode = os.stat(file_path).st_mode
+    try:
+        mode = os.stat(file_path).st_mode
+
```
Comment (ctb) [inline khmer/kfile.py]: unnecessary empty line ;)

File: tests/test_scripts.py [hunk 2]
```diff
@@ -1926,3 +1926,15 @@ def test_readstats_empty():
     assert status == 0
 
     assert expected_output in out
+
+
+def test_existance_failure():
```
Comment (ctb) [inline tests/test_scripts.py]: You didn't run a spellchecker :)

File: tests/test_scripts.py [hunk 3]
```diff
@@ -1926,3 +1926,15 @@ def test_readstats_empty():
...
+    expected_output = 'ERROR: Input file'
+
+    args = ['thisfiledoesnotexistatall']
```
Comment (ctb) [inline tests/test_scripts.py]: Please use:

utils.get_temp_filename('thisfiledoesnotexistatall')

File: tests/test_scripts.py [hunk 4]
```diff
@@ -1926,3 +1926,15 @@ def test_readstats_empty():
...
+
+    status, out, err = utils.runscript(
```
Comment (ctb) [inline tests/test_scripts.py]: This test should specifically test the changes you made to kfile.py - if you remove those changes, will this test still succeed? It doesn't look like it to me.

So, I think you should actually write two tests -- one test for a script like extract-paired-reads, that tests the error message and exit status.  And one test in test_script_arguments.py that runs the code in kfile directly, and verifies that behavior.  Note that sys.exit raises SystemExit, so you can look for that in the test.

(Let me know if you'd like me to expand on all of this!)

Comment (jessicamizzi) [inline tests/test_scripts.py]: Will the test in test_scripts.py work to as the first test you described?  If I'm understanding you, there only needs to be another test in test_script_arguments.py to test the code in kfile.  I would like some expansion on the test_script_arguments.py test if possible.

Also - how do I a run a spell checker? Thanks!

Comment (ctb) [inline tests/test_scripts.py]: for running a spell checker - what editor are you using? it'll depend on that. or you can just copy/paste the code into Word...

yes, I misread the test in test_scripts - it's good now!

the test in test_script_arguments should be of the specific kfile function, as opposed to running a script that runs the kfile function.  so, it should call kfile.check_file_status specifically, and verify that it raises a SystemExit exception when the file doesn't exist; and maybe another test to check that if you specify force, it doesn't exit; etc.  Basically you should test all of the if/else paths in the code.  Does that make more sense?

File: khmer/kfile.py [hunk 5]
```diff
@@ -22,7 +22,16 @@ def check_file_status(file_path, force):
...
+    except OSError as error:
```
Comment (mr-c) [inline khmer/kfile.py]: `W0612: (unused-variable), check_file_status: Unused variable 'error'` -> no need for `as error` here

File: tests/test_functions.py [hunk 6]
```diff
@@ -114,6 +114,21 @@ def test_extract_hashbits_info():
             print >>sys.stderr, '...failed to remove {fn}'.format(fn)
 
 
+def check_file_status_kfile():
+    try:
+        kfile.check_file_status('thisfiledoesnotexistatall')
```
Comment (ctb) [inline tests/test_functions.py]: put an 'assert False' after this, because it *must* raise an OSError to succeed.

File: tests/test_functions.py [hunk 7]
```diff
@@ -114,6 +114,21 @@ def test_extract_hashbits_info():
             print >>sys.stderr, '...failed to remove {fn}'.format(fn)
 
 
+def check_file_status_kfile():
+    try:
```
Comment (ctb) [inline tests/test_functions.py]: generate a guaranteed-not-to-exist filename with utils.get_temp_filename('thisfiledoesnotexist')

File: tests/test_functions.py [hunk 8]
```diff
@@ -114,6 +115,21 @@ def test_extract_hashbits_info():
...
+def test_check_file_status_kfile():
+    try:
+        utils.get_temp_filename('thisfiledoesnotexistatall')
+    except OSError as e:
+        print >>sys.stder, '...failed to remove {fn}'.format(fn)
+
+
+def test_check_file_status_kfile_force():
+    try:
+        khmer.kfile('thisfiledoesnotexistatall', force)
```
Comment (mr-c) [inline tests/test_functions.py]: `khmer.kfile` isn't the name of a function

File: tests/test_functions.py [hunk 9]
```diff
@@ -114,6 +115,24 @@ def test_extract_hashbits_info():
             print >>sys.stderr, '...failed to remove {fn}'.format(fn)
 
 
+def test_check_file_status_kfile():
+    fn = utils.get_temp_filename('thisfiledoesnotexist')
+    try:
+        check_file_status(fn, False)
+        assert True
```
Comment (jessicamizzi) [inline tests/test_functions.py]: @ctb could you remind me of the idea behind the if/if not statement to be put here you mentioned today?

Comment (ctb) [inline tests/test_functions.py]: You _could_ rewrite it as:

```
check_file_status_exited = False
try:
    check_file_status(fn, False)
except SystemExit:
    check_file_status_exited = True

assert check_file_status_exited
```

which makes the logic clear, at least.

**Comment by jessicamizzi** [comment]
@mr-c has the compatibility problem with pep8/autopep8 been fixed?

**Comment by mr-c** [comment]
Yes, sorry. Update to the latest from the master branch and you should be
good to go.

**Comment by mr-c** [comment]
Thanks!

**Pull request merged**